compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -242,29 +240,28 @@ impl<Prov> Allocation<Prov> {
     /// If `panic_on_fail` is true, this will never return `Err`.
     pub fn uninit<'tcx>(size: Size, align: Align, panic_on_fail: bool) -> InterpResult<'tcx, Self> {
         let align_usize: usize = align.bytes().try_into().unwrap();
-        let count_align = ((size.bytes_usize() / align_usize) + 1)*align_usize;
-
-        // TODO this one is supposed to handle allocation failure, so do so.
-        let mut vec_align: Vec<u8> = Vec::with_capacity(count_align);
-        vec_align.resize(count_align, 0);
-        let (buf, _len, _capacity) = vec_align.into_raw_parts();
-        vec_align = unsafe {Vec::from_raw_parts(buf as *mut u8, size.bytes_usize(), size.bytes_usize())};
+        let layout = std::alloc::Layout::from_size_align(size.bytes_usize(), align_usize).unwrap();
+        let vec_align = unsafe {
+            // https://doc.rust-lang.org/nightly/std/alloc/trait.GlobalAlloc.html#tymethod.alloc
+            // std::alloc::alloc returns null to indicate an allocation failure: 
+            // "Returning a null pointer indicates that either memory is exhausted 
+            // or layout does not meet this allocator’s size or alignment constraints."
+            let buf = std::alloc::alloc(layout);
+            // Handle allocation failure.
+            if buf.is_null() {
+                if panic_on_fail {
+                    panic!("Allocation::uninit called with panic_on_fail had allocation failure")
+                }
+                ty::tls::with(|tcx| {
+                    tcx.sess.delay_span_bug(DUMMY_SP, "exhausted memory during interpretation")
+                });
+                Err(InterpError::ResourceExhaustion(ResourceExhaustionInfo::MemoryExhausted))?
+            } 
+            Vec::from_raw_parts(buf as *mut u8, size.bytes_usize(), size.bytes_usize())
```

Comment:
You're not zeroing this, so this is UB. The buffer needs to be zeroed before it is converted to a slice or vector that doesn't have `MaybeUninit` - this can either be done by converting it to a type using `MaybeUninit` and then using `.assume_init()` after zeroing, or be done by zeroing it while it's still a pointer.